elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/expression/JoinReferenceExtractor.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2021, Yahoo Inc.
+ * Licensed under the Apache License, Version 2.0
+ * See LICENSE file in project root for terms.
+ */
+
+package com.yahoo.elide.datastores.aggregation.queryengines.sql.expression;
+
+import static com.yahoo.elide.core.utils.TypeHelper.appendAlias;
+import static com.yahoo.elide.datastores.aggregation.queryengines.sql.metadata.SQLReferenceTable.applyQuotes;
+import static com.yahoo.elide.datastores.aggregation.queryengines.sql.metadata.SQLReferenceTable.hasSql;
+import static com.yahoo.elide.datastores.aggregation.queryengines.sql.metadata.SQLReferenceTable.resolveTableOrSubselect;
+import static org.apache.commons.lang3.StringUtils.EMPTY;
+
+import com.yahoo.elide.core.Path.PathElement;
+import com.yahoo.elide.core.dictionary.EntityDictionary;
+import com.yahoo.elide.core.request.Argument;
+import com.yahoo.elide.core.type.Type;
+import com.yahoo.elide.datastores.aggregation.annotation.JoinType;
+import com.yahoo.elide.datastores.aggregation.core.JoinPath;
+import com.yahoo.elide.datastores.aggregation.metadata.MetaDataStore;
+import com.yahoo.elide.datastores.aggregation.metadata.TableContext;
+import com.yahoo.elide.datastores.aggregation.queryengines.sql.metadata.SQLJoin;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Provides Join Expressions for all the {@link JoinReference} in a given reference tree.
+ */
+public class JoinReferenceExtractor implements ReferenceVisitor<Set<String>> {
+
+    private static final String ON = "ON ";
+    private static final String OPEN_BRACKET = "(";
+    private static final String CLOSE_BRACKET = ")";
+
+    private final Set<String> joinExpressions;
+    private final TableContext tableCtx;
+    private final Map<String, Argument> columnArgs;
+    private final MetaDataStore metaDataStore;
+    private final EntityDictionary dictionary;
+
+    public JoinReferenceExtractor(TableContext tableCtx, Map<String, Argument> columnArgs,
+                    Set<String> joinExpressions) {
```

Comment:
It might be cleaner here to not pass in joinExpressions as an argument and always have this visitor build its own Set.  You can set union the sets later when you need to combine them.